src/ema/tests.rs

```diff
@@ -0,0 +1,98 @@
+use super::*;
+
+use crate::types::Price;
+
+use num_traits::{Bounded, One, Zero};
+use sp_arithmetic::{traits::Saturating, FixedPointNumber, FixedU128};
+
+#[test]
+fn ema_stays_stable_if_the_value_does_not_change() {
+    let alpha = smoothing_from_period(7);
+    debug_assert!(alpha <= Price::one());
+
+    let start_price = Price::saturating_from_integer(4u32);
+    let incoming_price = start_price;
+    let next_price = price_ema(start_price, incoming_price, alpha);
+    assert_eq!(next_price, start_price);
+    let start_balance = 4u32.into();
+    let incoming_balance = start_balance;
+    let next_balance = balance_ema(start_balance, incoming_balance, alpha);
+    assert_eq!(next_balance, start_balance);
+}
+
+#[test]
+fn ema_works() {
+    let alpha = smoothing_from_period(7);
+    debug_assert!(alpha <= Price::one());
+
+    // price
+    let start_price = 4.into();
+    let incoming_price = 8.into();
+    let next_price = price_ema(start_price, incoming_price, alpha);
+    assert_eq!(next_price, 5.into());
+
+    let start_price = Price::saturating_from_rational(4, 100);
+    let incoming_price = Price::saturating_from_rational(8, 100);
+    let next_price = price_ema(start_price, incoming_price, alpha);
+    assert_eq!(next_price, Price::saturating_from_rational(5, 100));
+
+    // balance
+    let start_balance = 4u128;
+    let incoming_balance = 8u128;
+    let next_balance = balance_ema(start_balance, incoming_balance, alpha);
+    assert_eq!(next_balance, 5u128);
+
+    // volume
+    let start_volume = (4u128, 1u128, 8u128, 0u128);
+    let incoming_volume = (8u128, 1u128, 4u128, 0u128);
+    let next_volume = volume_ema(start_volume, incoming_volume, alpha);
+    assert_eq!(next_volume, (5u128, 1u128, 7u128, 0u128));
+}
+
+#[test]
+fn price_ema_boundary_values() {
+    let alpha = Price::saturating_from_rational(1, 2);
+    debug_assert!(alpha <= Price::one());
+
+    // previously zero, incoming max
+    let next_price = price_ema(Price::zero(), Price::max_value(), alpha);
+    assert_eq!(next_price, Price::max_value() / 2.into());
+    // the oracle is biased towards the previous value
+    let bias = Price::saturating_from_rational(1, Price::DIV);
+    // previously max, incoming zero
+    let next_price = price_ema(Price::max_value(), Price::zero(), alpha);
+    assert_eq!(next_price, Price::max_value() / 2.into() + bias);
+}
+
+#[test]
+fn ema_does_not_saturate_on_big_balances() {
+    let alpha = Price::one();
+
+    let start_balance = u128::MAX;
+    let incoming_balance = u128::MAX;
+    let next_balance = balance_ema(start_balance, incoming_balance, alpha);
+    assert_eq!(next_balance, incoming_balance);
+}
+
+#[test]
+fn exp_smoothing_works() {
+    let smoothing = smoothing_from_period(7);
```

Comment:
7 is not a common value (that would be 100, the ten minutes, or 14_400, the daily, oracle), but it is a possible value that should work and has the benefit of being easy to test

Reply:
K, that's fair for an example-based test. But I think by using rug crate to test against arbitrarily accurate math, we can test more realistic value ranges without worrying about what values allow for manual calculation.

Reply:
there should be property tests addressing this (at least by now), do you think it should remove the simple example test? I think it might be better to have than not have